cleverhans/model_zoo/deep_k_nearest_neighbors/dknn.py

```diff
@@ -0,0 +1,429 @@
+    """
+    Given an array of nb_data x nb_classes dimensions, use conformal prediction to compute
+    the DkNN's prediction, confidence and credibility.
+    """
+    nb_data = knns_not_in_class.shape[0]
+    preds_knn = np.zeros(nb_data, dtype=np.int32)
+    confs = np.zeros((nb_data, self.nb_classes), dtype=np.float32)
+    creds = np.zeros((nb_data, self.nb_classes), dtype=np.float32)
+
+    for i in range(nb_data):
+      # p-value of test input for each class
+      p_value = np.zeros(self.nb_classes, dtype=np.float32)
+
+      for class_id in range(self.nb_classes):
+        # p-value of (test point, candidate label)
+        p_value[class_id] = (float(self.nb_cali) - bisect_left(
+            self.cali_nonconformity, knns_not_in_class[i, class_id])) / float(self.nb_cali)
+
+      preds_knn[i] = np.argmax(p_value)
+      confs[i, preds_knn[i]] = 1. - p_value[np.argsort(p_value)[-2]]
+      creds[i, preds_knn[i]] = p_value[preds_knn[i]]
+    return preds_knn, confs, creds
+
+  def fprop_np(self, data_np):
+    """
+    Performs a forward pass through the DkNN on an numpy array of data.
+    """
+    if not self.calibrated:
+      raise ValueError(
+          "DkNN needs to be calibrated by calling DkNNModel.calibrate method once before inferring.")
+    data_activations = self.get_activations(data_np)
+    _, knns_labels = self.find_train_knns(data_activations)
+    knns_not_in_class = self.nonconformity(knns_labels)
+    _, _, creds = self.preds_conf_cred(knns_not_in_class)
+    return creds
+
+  def fprop(self, x):
+    """
+    Performs a forward pass through the DkNN on a TF tensor by wrapping
+    the fprop_np method.
+    """
+    logits = tf.py_func(self.fprop_np, [x], tf.float32)
+    return {self.O_LOGITS: logits}
+
+  def calibrate(self, cali_data, cali_labels):
+    """
+    Runs the DkNN on holdout data to calibrate the credibility metric.
+    :param cali_data: np array of calibration data.
+    :param cali_labels: np vector of calibration labels.
+    """
+    self.nb_cali = cali_labels.shape[0]
+    self.cali_activations = self.get_activations(cali_data)
+    self.cali_labels = cali_labels
+
+    print("Starting calibration of DkNN.")
+    cali_knns_ind, cali_knns_labels = self.find_train_knns(
+        self.cali_activations)
+    assert all([v.shape == (self.nb_cali, self.neighbors)
+                for v in cali_knns_ind.itervalues()])
+    assert all([v.shape == (self.nb_cali, self.neighbors)
+                for v in cali_knns_labels.itervalues()])
+
+    cali_knns_not_in_class = self.nonconformity(cali_knns_labels)
+    cali_knns_not_in_l = np.zeros(self.nb_cali, dtype=np.int32)
+    for i in range(self.nb_cali):
+      cali_knns_not_in_l[i] = cali_knns_not_in_class[i, cali_labels[i]]
+    cali_knns_not_in_l_sorted = np.sort(cali_knns_not_in_l)
+    try:
+      last_zero = int(np.argwhere(cali_knns_not_in_l_sorted == 0)[-1])
+    except:  # pylint: disable-msg=W0702
+      last_zero = 0
```

Comment:
How is this distinguishable from the case where the last 0 is actually at index 0 in cali_knns_not_in_l_sorted?

Reply:
I replaced this with a call to `np.trim_zeros`, which is probably more readable.